### DISCUSSION TITLE
RTU Server works by mbpoll but not with pymodbus

### DISCUSSION THREAD

**Discussion by srolf** [pymodbus-dev/pymodbus#1651] [Help] [answered]
Hi all, I'm pretty new to Modbus and pymodbus.

Problem: A self-created RTU server providing a switch status (ESP32 with [modbus-serial](https://github.com/epsilonrt/modbus-serial) lib) can be accessed via [mbpoll](https://github.com/epsilonrt/mbpoll) but not via pymodbus (Raspberry Pi2 with a RS485 HAT).

Running mbpoll on RPi:
```
$ mbpoll -m rtu -b38400 -P none -a13 -t1 /dev/serial0 -v
debug enabled
Set device=/dev/serial0
mbpoll 1.0-0 - FieldTalk(tm) Modbus(R) Master Simulator
Copyright © 2015-2019 Pascal JEAN, https://github.com/epsilonrt/mbpoll
This program comes with ABSOLUTELY NO WARRANTY.
This is free software, and you are welcome to redistribute it
under certain conditions; type 'mbpoll -w' for details.

Opening /dev/serial0 at 38400 bauds (N, 8, 1)
Set response timeout to 1 sec, 0 us
Protocol configuration: Modbus RTU
Slave configuration...: address = [13]
                        start reference = 1, count = 1
Communication.........: /dev/serial0,      38400-8N1 
                        t/o 1.00 s, poll rate 1000 ms
Data type.............: discrete input

-- Polling slave 13... Ctrl-C to stop)
[0D][02][00][00][00][01][B9][06]
Waiting for a confirmation...
<0D><02><01><00><A2><D8>
[1]: 	0
-- Polling slave 13... Ctrl-C to stop)
[0D][02][00][00][00][01][B9][06]
Waiting for a confirmation...
<0D><02><01><01><63><18>
[1]: 	1
-- Polling slave 13... Ctrl-C to stop)
[0D][02][00][00][00][01][B9][06]
Waiting for a confirmation...
<0D><02><01><01><63><18>
[1]: 	1
-- Polling slave 13... Ctrl-C to stop)
[0D][02][00][00][00][01][B9][06]
Waiting for a confirmation...
<0D><02><01><01><63><18>
[1]: 	1
-- Polling slave 13... Ctrl-C to stop)
[0D][02][00][00][00][01][B9][06]
Waiting for a confirmation...
<0D><02><01><00><A2><D8>
[1]: 	0
-- Polling slave 13... Ctrl-C to stop)
[0D][02][00][00][00][01][B9][06]
Waiting for a confirmation...
<0D><02><01><00><A2><D8>
[1]: 	0
^C--- /dev/serial0 poll statistics ---
6 frames transmitted, 6 received, 0 errors, 0.0% frame loss
```
The pymodbus RTU client implementation looks like this:
```
import pymodbus
import serial
from pymodbus.pdu import ModbusRequest
from pymodbus.client import ModbusSerialClient as ModbusClient #initialize a serial RTU client instance
from pymodbus.transaction import ModbusRtuFramer

client=ModbusClient(method="rtu", port="/dev/serial0", stopbits=1, bytesize=8, parity='N', baudrate=38400)

#Connect to the serial modbus server
connection = client.connect()
print (connection)

#Starting add, num of reg to read, slave unit.
result=client.read_discrete_inputs(address=1, count=1, slave=13)

print(result)

#Closes the underlying connection
client.close()
```
The response looks like this:
```
True
Modbus Error: [Input/Output] No Response received from the remote slave/Unable to decode response
```
Any idea about what I'm doing wrong?

**Comment by srolf** [comment]
The upper implementation didn't include logging, so I've updated the python script to:
```
import pymodbus
import serial
from pymodbus.pdu import ModbusRequest
from pymodbus.client import ModbusSerialClient as ModbusClient #initialize a serial RTU client instance
from pymodbus.transaction import ModbusRtuFramer

import logging
logging.basicConfig()
log = logging.getLogger('pymodbus')
log.setLevel(logging.DEBUG)

client=ModbusClient(method="rtu", port="/dev/serial0", stopbits=1, bytesize=8, parity='N', baudrate=38400)

#Connect to the serial modbus server
connection = client.connect()
print (connection)

#Starting add, num of reg to read, slave unit.
result=client.read_discrete_inputs(address=1, count=1, slave=13)

print(result)

#Closes the underlying connection
client.close()
```
The log out looks like this:
```
True
DEBUG:pymodbus.logging:Current transaction state - IDLE
DEBUG:pymodbus.logging:Running transaction 1
DEBUG:pymodbus.logging:SEND: 0xd 0x2 0x0 0x1 0x0 0x1 0xe8 0xc6
DEBUG:pymodbus.logging:New Transaction state "SENDING"
DEBUG:pymodbus.logging:Changing transaction state from "SENDING" to "WAITING FOR REPLY"
DEBUG:pymodbus.logging:Changing transaction state from "WAITING FOR REPLY" to "PROCESSING REPLY"
DEBUG:pymodbus.logging:RECV: 0xd 0x82 0x2 0x1 0x62
DEBUG:pymodbus.logging:Getting transaction 13
DEBUG:pymodbus.logging:Changing transaction state from "PROCESSING REPLY" to "TRANSACTION_COMPLETE"
Modbus Error: [Input/Output] No Response received from the remote slave/Unable to decode response
```

**Comment by janiversen** [comment]
please add the debug log as pr documentation, call pymodbus_apply.....

**Comment by janiversen** [comment]
Please also write which version you are using, hardware etc.

**Reply by janiversen** [reply]
Did you check that you use the correct framer, that is often the problem, again see documentation

**Reply by srolf** [reply]
Thank you very much for having a look at my issue.

I made some progress in the evening, but let me first complete the requested info.

On Raspberry Pi 2 B:
- Ubuntu 22.04.2 LTS Server armv7l
- pymodbus 3.3.2

On ESP32:
- Ardunio ESP32 2.0.9
- Modbus-Arduino 1.3.0
- Modbus-Serial 2.0.6

Meanwhile I'm able to fetch the discrete input register value. I caused the trouble by using an **incorrect address: not 1, but 0 is correct**. Maybe I used 1 because of the way mbpoll output looks like.

And for sure, I will also have a closer look how the logging shall be done for any new questions.
```
#result=client.read_discrete_inputs(address=1, count=1, slave=13)
result=client.read_discrete_inputs(address=0, count=1, slave=13)

print(result)
print(result.encode())
print(result.encode()[1])
```
But I'm still slightly confused how the output of the 3 print statements looks like - seems I need to read way more about Modbus and pymodbus. Maybe you could give me some hints:

```
ReadDiscreteInputsResponse(8)
b'\x01\x00'
0
```
The first one is for me looking uncommon: Does it point to a payload of 8 bit == 1 byte?
The second one also irritates myself little, is the 1st byte pointing to the length in bytes of the payload and the 2nd byte is the payload?

**Reply by janiversen** [reply]
It is much easier if you read the documentation, encode() is an internal function.

**Reply by janiversen** [reply]
Just a little thing, why do you encode a result, I really do not see the sense of that....normally you encode a request and decode a response.

**Reply by srolf** [reply]
As I tried to explain, the way the response looks like is confusing me. While looking into the pymodbus documentation, I was not able to find the definition of `ModbusResponse` as returned by [`read_discrete_inputs`](https://pymodbus.readthedocs.io/en/dev/source/library/client.html#pymodbus.client.mixin.ModbusClientMixin.read_discrete_inputs).

If I use in the code
`print(result.bits)`
and the switch on the ESP32 is closed during request,
the presentation of it looks way more straight:
`[True, False, False, False, False, False, False, False]`
But I derived it only from the the code of given examples in the documentation.

**Reply by janiversen** [reply, marked as answer]
Which is why examples are part of the documentation. The examples contains e.g. a template call showing how to to look for error conditions. All modbus calls are shown in one example, so I am not sure what you want more.

**Reply by srolf** [reply]
Please don't get me wrong: Looking into the examples is fine. Covering also ModbusResponse in the API documentation might make the start with pymodbus for newbies like me little easier.

**Reply by janiversen** [reply]
Pull requests are more than welcome.
